pystiche_papers/sanakoyeu_et_al_2018/_modules.py

```diff
@@ -184,3 +192,202 @@ def residual_block(channels: int) -> ResidualBlock:
+        kernel_size: Union[Tuple[int, int], int] = 10,
+        stride: Union[Tuple[int, int], int] = 1,
+        padding: str = "same",
+        impl_params: bool = True,
+    ):
+        super().__init__()
+        self.impl_params = impl_params
+
+        padding = get_padding(padding, kernel_size)
+
+        self.forwardBlock = get_transformation_block(
+            in_channels, kernel_size, stride, padding, impl_params=impl_params,
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        if self.impl_params:
+            return cast(torch.Tensor, self.forwardBlock(input))
+        else:
+            return cast(torch.Tensor, nn.utils.weight_norm(self.forwardBlock(input)))
+
+
+def discriminator_encoder_modules(
+    in_channels: int = 3, inplace: bool = True,
+) -> Dict[str, nn.Sequential]:
+    # FIXME:  if/when the Python interpreter will learn to accept the correct signature
+    # https://stackoverflow.com/questions/41207128/how-do-i-specify-ordereddict-k-v-types-for-mypy-type-annotation
+
+    modules = OrderedDict(
+        {
+            "scale_0": ConvBlock(
+                in_channels,
+                128,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_1": ConvBlock(
+                128,
+                128,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_2": ConvBlock(
+                128,
+                256,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_3": ConvBlock(
+                256,
+                512,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_4": ConvBlock(
+                in_channels=512,
+                out_channels=512,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_5": ConvBlock(
+                512,
+                1024,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_6": ConvBlock(
+                1024,
+                1024,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+        }
+    )
+    return cast(Dict[str, nn.Sequential], modules)
+
+
+class DiscriminatorEncoder(MultiLayerEncoder):
```

Comment:
If there is an "encoder part" shouldn't there also be a `Decoder`?

Reply:
No a `decoder` is not needed here. This `encoder `is a `multi_layer_encoder `for a modified `MultiLayerEncodingOperator`, since the `discriminator` structure has a forward part, from which the encodings are evaluated on predefined layers using the additional `prediction_modules`. I should probably rename this to a `discriminator_multi_layer_encoder` though.

Reply:
Maybe we should split this into two entitites: `Discriminator` and `DiscriminatorMultiLayerEncoder`. Although we won't use the `Discriminator` for our implementation (other than as base for the `DiscriminatorMultiLayerEncoder`) I think it makes the distinction more clear.

Reply:
The complete `Discriminator `is defined as a `MultiLayerEncodingOperator `in loss, if that's what you mean.

Reply:
> The complete `Discriminator` is defined as a `MultiLayerEncodingOperator` in loss, if that's what you mean.

That doesn't make sense. Do you mean: "The `Discriminator` servers as `MultiLayerEncoder` within a `MultiLayerEncodingOperator`"?

---

What I mean is that we should have a simple sequential `Discriminator` as the authors describe in their paper. For our implementation (to match the reference implementation) we also need an `MultiLayerDiscriminator`. I think we should keep both separate to avoid confusion.

Reply:
Yes, i mean the `Discriminator` serves as `MultiLayerEncoder` within a `MultiLayerEncodingOperator`.